### PR TITLE
Updated (And Simplified) Pen Grouping

### PR DESCRIPTION
**Pull request by cvankerkhove**
This pull request is to update the previously merged grouping_algo branch. The files should be up to standards with the commenting, variable labeling, and function descriptions. There is a significant amount of unnecessary code cut out by using just a nutrition requirement percentile grouping. Furthermore, the method allocates the cows with higher nutritional requirements first (into pens with lower ID numbers). The file should also run seamlessly with any number of pen inputs in the json file.

### CODE REVIEW

**Review by militsasotirova — APPROVED** [review]
I like the organization/documentation that you've added. Both animal_management.json and barnyard.json inputs look like they work! 
Animal_management.json does take an additional minute to run now - but I'm assuming that's coming from all of the calculations/stats that are added?

File: RUFAS/routines/animal/animal_management.py [hunk 1]
```diff
@@ -224,18 +225,18 @@ def pen_allocation(self):
...
         '''
-
+        #Assiging non-cows to pens
         if len(self.all_pens) == 3:
             self.all_pens[0].update_animals(self.calves)
         elif len(self.all_pens) == 4:
             heifers = self.heiferIs + self.heiferIIs + self.heiferIIIs
             self.all_pens[0].update_animals(self.calves)
             self.all_pens[1].update_animals(heifers)
         elif len(self.all_pens) == 5:
```
Comment (KFosterReed) [inline RUFAS/routines/animal/animal_management.py:232]: I think this if statement double assigns heiferIIs to a pen and leaves out heiferIs

Comment (KFosterReed) [reply RUFAS/routines/animal/animal_management.py:232]: I think what was intended (and what makes more sense biologically/logistically) is for L238 to be the pen allocation for heifer1

File: RUFAS/routines/animal/animal_management.py [hunk 2]
```diff
@@ -224,18 +225,18 @@ def pen_allocation(self):
...
-            heifers = self.heiferIs + self.heiferIIs
+            heifers = self.heiferIIs + self.heiferIIIs
             self.all_pens[0].update_animals(self.calves)
-            self.all_pens[1].update_animals(heifers)
-            self.all_pens[2].update_animals(self.heiferIIIs)
+            self.all_pens[1].update_animals(self.heiferIIs)
```
Comment (KFosterReed) [inline RUFAS/routines/animal/animal_management.py]:
```suggestion
            self.all_pens[1].update_animals(self.heiferIs)
```

File: RUFAS/routines/animal/animal_management.py [hunk 3]
```diff
@@ -251,285 +252,26 @@ def pen_allocation(self):
                 lactating_cows.append(cow)
             else:
                 dry_cows.append(cow)
+        #Asigning Dry Cows to Pens
         if len(self.all_pens) == 3:
             dry_and_heifers = self.heiferIs + self.heiferIIs + self.heiferIIIs + dry_cows
             self.all_pens[1].update_animals(dry_and_heifers)
             self.all_pens[2].update_animals(lactating_cows)
-        elif len(self.all_pens) == 4:
+        elif (4 <= len(self.all_pens) <= 6):
```
Comment (KFosterReed) [inline RUFAS/routines/animal/animal_management.py:257]: This makes sense to me if the number of pens is 4 but if pens is 5 or 6, doesn't this double assign pens and leave some pens open?

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 4]
```diff
@@ -2,10 +2,11 @@
 '''
 RUFAS: Ruminant Farm Systems Model
 File name: clustering_pen_grouping.py
-Description: This file's main function is grouping(list) (line 54) which returns
-    a 2D list with each list in the matrix being a pen grouping of cows based on
-    nutritional requirments. This function is called in animal_management.py for
-    each ration cycle if there are more than 7 pens in the input.
+Description: This file's main function is grouping(list, pens) (line 44) which returns
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py:4]: Great description! :)

File: RUFAS/routines/animal/clustering_pen_grouping.py [hunk 5]
```diff
@@ -40,265 +41,123 @@ def percentile_list(l):
     return(percentile_list)
 
 
-def grouping(list):
+def grouping(list, pens):
     '''Grouping algorithim that utilizes k-means clustering and takes an input
        that is a list of objects of class Cow (see cow.py) and groups them into
        exactly 1 of 14 different pens. This function returns a list of 14 lists
```
Comment (KFosterReed) [inline RUFAS/routines/animal/clustering_pen_grouping.py]: I think the number 14 corresponds to the Jorge's original method? I suggest change to 'groups them into 2 or more pens depending on the total number of pens. This function returns a list of cow groupings (lists of objects of class Cow).

**Comment by cvankerkhove** [comment]
> I like the organization/documentation that you've added. Both animal_management.json and barnyard.json inputs look like they work!
> Animal_management.json does take an additional minute to run now - but I'm assuming that's coming from all of the calculations/stats that are added?

**Comment by KFosterReed** [comment]
So it all worked really well until I tried to run it without any pens. The initial warning message popped up so that was a good sign but then I got a divide by zero error because the DMIest in the nutrient requirement calculations was zero. I spent a while trying to debug this but was unsuccessful. Any thoughts on what is going on?

**Comment by KFosterReed** [comment]
So I may have found a solution but I want to check with @cvankerkhove  and @militsasotirova to see if it is correct use of python data types. Based on our conversation yesterday, I looked at what was using pen_information from the .json data dictionary. In animal_management.py init_animals() directly after init_pen() uses pen_information and only calls init_nutrient_rqmts() if the pen_count >0. I changed the input for Pen_data in init_animals() from data['pen_information'] to self.all_pens. Is this an OK way to pass this information to the init_animals function?

I tried to put a comment at the couple of lines that I changed but it wouldn't let me. So my suggestion is to change L83 in animal_management.py to: 

        self.init_animals(data['herd_information'], self.all_pens, feed)
        #self.init_animals(data['herd_information'], data['pen_information'], feed)

And change L171 to: 
        pen_count = len(Pen_data)
        #pen_count = 0
        #for pen in Pen_data:
           #pen_count += 1

The old code is shown as commented out above.

**Comment by militsasotirova** [comment]
> So I may have found a solution but I want to check with @cvankerkhove and @militsasotirova to see if it is correct use of python data types. Based on our conversation yesterday, I looked at what was using pen_information from the .json data dictionary. In animal_management.py init_animals() directly after init_pen() uses pen_information and only calls init_nutrient_rqmts() if the pen_count >0. I changed the input for Pen_data in init_animals() from data['pen_information'] to self.all_pens. Is this an OK way to pass this information to the init_animals function?

That sounds like a good idea! In fact, instead of passing in Pen_data, we could just have the following if statement 
if len(self.all_pens) > 0:
since we just use the count in init_animals(). 

Also, I'm just noticing this now, but Pen_data and Herd_data should be lowercase to follow Python conventions.

**Comment by KFosterReed** [comment]
> > So I may have found a solution but I want to check with @cvankerkhove and @militsasotirova to see if it is correct use of python data types. Based on our conversation yesterday, I looked at what was using pen_information from the .json data dictionary. In animal_management.py init_animals() directly after init_pen() uses pen_information and only calls init_nutrient_rqmts() if the pen_count >0. I changed the input for Pen_data in init_animals() from data['pen_information'] to self.all_pens. Is this an OK way to pass this information to the init_animals function?
> 
> That sounds like a good idea! In fact, instead of passing in Pen_data, we could just pass in len(self.all_pens) since we just use the count in init_animals().
> 
> Also, I'm just noticing this now, but Pen_data and Herd_data should be lowercase to follow Python conventions.

Good catch on the case convention in the names:) @cvankerkhove does this sound like a plan?

**Review by KFosterReed — APPROVED** [review]
Awesome! This now works with any number of pens. Great work team!

**Pull request merged**